### DISCUSSION TITLE
"Last GCD Move"

### DISCUSSION THREAD

**Discussion by Evolutious** [PrincessRTFM/XIVComboPlugin#50] [General]
Hi @PrincessRTFM - I am working on some stuff for SAM and there is one action, Tsubame-gaeshi, that becomes active directly after using Iaijutsu, but it has to be your next GCD, meaning that if you use Iaijutsu then use another GCD other than Tsubame-gaeshi, then Tsubame-gaeshi is no longer usable.

Is there a method in your fork that can do a check to determine your previously used GCD action? Unfortunately, you are not able to use any of the standard things to leverage for this, such as having a certain buff or using a LastComboMove; none of those apply to this.

**Comment by VariableVixen** [comment]
To the best of my knowledge, there's nothing _yet_ in place for this. Clearly the game itself has some mechanism to determine that, but it hasn't been hooked or duplicated anywhere that I'm aware of. Presumably there's either a single value stored somewhere (in which case you'd just need to read it, but the hard part would be finding it, and especially finding a signature to look for rather than using an offset) or there's a function call somewhere that returns a boolean (which would be easier to find and call) but I've never seen anything for that.

My best suggestion is to check other forks of the original and see if any of them have tried to implement such a feature; if you can find it in someone else's version, you could use that as a basis for your own.

My _worst_ suggestion would be to hook into the player performing an action in order to cache things yourself, and try to reverse-engineer the precise conditions the game uses to duplicate the checks manually. It'd _work_, but that's about the only good thing I can say about that approach.

**Reply by Evolutious** [reply]
i will look around and see what I can find. I don't see anything under LocalPlayer that would provide any information about "last used action" or the sort, so don't think that will work. I was thinking about making a public variable that i can just turn off/on based on action used. For example, if i use Iaijutsu, then i flip this bool true, then when i use Tsubame-gaeshi i flip it back false. then all i would ahve to do for Tsubame-gaeshi is check that bool if it is true.

My issue with that approach is i have no idea how to make such a public variable (I am a noob to C#). What do you think of this approach? Do you know how to do this by chance?

**Reply by VariableVixen** [reply]
That'd be real easy, but it also wouldn't work. You'd have to set it to false when _any_ action except Tsubame-gaeshi is used:

> if you use Iaijutsu then use another GCD other than Tsubame-gaeshi, then Tsubame-gaeshi is no longer usable

Also, you might want to look into some basic C# tutorials; methods, fields, and properties are pretty much core parts of the language and if you want to write C# then you should really know them, and at least the most common modifiers like [access levels](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/access-modifiers), `static`, and `readonly`. Specifically, you'd probably want a signature of `internal [static] bool`, but I don't know where you'd put it - maybe in the caching class.

**Reply by Evolutious** [reply]
Adding a flip to false to every action will not be an issue for what I am working on ;)

I will do some research on what you provided. Thanks!

**Reply by Evolutious** [reply]
Hey @PrincessRTFM - Just letting you know that either this was recently updated to support or I was just overlooking it, but you can use the following to determine if any of the Kaeshi actions are ready to be used:

~~~C#
gauge.Kaeshi == Kaeshi.NAMIKIRI

gauge.Kaeshi == Kaeshi.HIGANBANA
gauge.Kaeshi == Kaeshi.GOKEN
gauge.Kaeshi == Kaeshi.SETSUGEKKA
~~~

So in my code, for single target, I used:

~~~C#
// Tsubame-gaeshi
if (level >= SAM.Levels.TsubameGaeshi && (gauge.Kaeshi == Kaeshi.SETSUGEKKA || gauge.Kaeshi == Kaeshi.GOKEN) && IsOffCooldown(SAM.TsubameGaeshi)) { return OriginalHook(SAM.TsubameGaeshi); }
~~~
